### ISSUE TITLE
download game data percentage

### ISSUE THREAD

**Issue by shaylevary** [emulatorjs/emulatorjs#332]
Hi,

I've noticed today that the "download game data" percentage doesn't appear on the latest versions.

Instead of showing percentage like this:
![Screenshot_20220815-221345_Chrome](https://user-images.githubusercontent.com/10597517/184702489-6d9baa63-0bf4-4818-bda3-e6751234ce56.jpg)

It shows like this:
![Screenshot_20220815-221459_Chrome](https://user-images.githubusercontent.com/10597517/184702552-9422848b-7ff2-48d8-9f00-f556bbf3f48b.jpg)

**Comment by ethanaobrien** [comment]
Well I took that out because it was unstable with the new request function I wrote


Please give the issues a break you don't need to open an issue for every little problem

**Comment by shaylevary** [comment]
I would appreciate if you could please find the time to re-update the "download game data percentage" function, because most PSX/3DO ROMs files are 200mb+, so it will be more clear and nice UI indication for the user to know how much time left to finish download the rom and start the game.
Currently, users may think that the script is stuck or the game isn't working - when the "Loading Game Data message" appears without any indication that the file is still being loaded, while sometimes it can take several minutes (depending on the amount of Internet traffic).

BTW, Is there any option to set download rom caching? so bigger rom files will be download only once, and be saved on user's temp dir for X time?

**Comment by ethanaobrien** [comment]
That's a good point, I'll do this when I get a chance. There's no way to only cache big roms (although it caches roms under a certain size, not sure why) I'd be happy to either make an option to only cache big roms (> 20mb) or to take out the rom size restriction. Let me know what would work better for you.

**Comment by Ramaerel** [comment]
> BTW, Is there any option to set download rom caching? so bigger rom files will be download only once, and be saved on user's temp dir for X time?

I don't remember if this got implemented or not, but I recall something about using a blob in place of a url for the rom path. If this was actually implemented, you could probably build this yourself using something that stores locally like IndexDB or something. Just an idea

Edit to just say I didn't see Ethan's response 😶

**Comment by ethanaobrien** [comment]
It's all good, yes you could do that too. Blob urls are supported and will not be cached. Theoretically you can set the game url to a blob/arraybuffer/uint8array (assuming I didn't break that) but keep in mind when using a blob or other data type, that you must specify the gameName or the save files will not be saved in the browser and the save state will be named game.state

**Comment by shaylevary** [comment]
> That's a good point, I'll do this when I get a chance. There's no way to only cache big roms (although it caches roms under a certain size, not sure why) I'd be happy to either make an option to only cache big roms (> 20mb) or to take out the rom size restriction. Let me know what would work better for you.

In my opinion, the best solution would be to set the maximum filesize cache limit through the script js code.
for example:
**EJS_cachelimit: 50**
* Files that exceed the size of 50 megabytes won't be cached.

BTW, Is there a way to clear those cached files after a period of time? Or will the user have to clear his browser cache eachtime? Another alternative solution, if possible is to add a button of "clear cache" on the bottom of the script interface.

**Comment by ethanaobrien** [comment]
I can add the cache limit thing


There's a hidden button you can enable to manage the cached rom files

```
EJS_Buttons = {
    cacheManage: true
}
```

You can enable it with the code above

**Comment by shaylevary** [comment]
I've already tried to enable this button, but it doesn't seems to appear on the interface buttons.


![cache_button](https://user-images.githubusercontent.com/10597517/186416403-dddb632f-d820-4928-b5e3-78a43451615c.jpg)

**Comment by ethanaobrien** [comment]
It's located in the right-click context menu

**Comment by shaylevary** [comment]
Yes, you are right, i didn't notice this feature appeae inside the right-click menu (i still would prefer to see a buttton with this feature on the bottom menu - with an option to enable/disable it).


Anyway, i've tried to click on the "remove" button, but it doesn't work.
Here is my console log when i click on "remove":

ncaught ReferenceError: _0x277c67 is not defined
    at i.querySelector.onclick (emulator.min.js?v=23.7:1:287654)
i.querySelector.onclick @ emulator.min.js?v=23.7:1

**Comment by shaylevary** [comment]
I would like to add that it will be nice to have "remove all" button on top of the list, if user wants to clear all the games in cache instead of click "remove" in each one separately.

**Comment by ethanaobrien** [comment]
I'm probably not going to add that. Sorry

**Comment by ethanaobrien** [comment]
Fixed

**Comment by shaylevary** [comment]
Thank you for this fix, I really appreciate your hard work.
Btw, Did you add the time to add the cache limit feature on this build?

**Comment by ethanaobrien** [comment]
Actually, I completely took out the limit.

**Comment by shaylevary** [comment]
Hi,
I've download the latest build, but it seems that the "download game data percentage" still doesn't appear (i've tried to test it on psx, 3do, mame2003 roms).

The second issue was the cache file manager, it seems to work ok now when i click on the remove button, but how do i get to the right click menu in order to enter the cache file manager from my mobile? i think that the best solution is to add the "cache manager" button on the bottom menu as i've suggested.

And third issue was about remove the cache limit, i've tried to test it with bigger psx files (90mb+), the rom file appears under the cache manager, but when refresh the page, it still download the rom file regulary from server, and not load it immediately from cache.

Thank you again for your hard work!

**Comment by ethanaobrien** [comment]
The download game percentage does appear, it might not appear if 1. the server doesnt set the `content-length` header or if 2. the game url is not https or http.

I'm not adding the button to the bottom, its better in the right click menu.

I tested, it caches the roms for me. I have no clue how you're experiencing these things.

You know you could actually learn to improve the code rather than just making issues so you'd actually help fix them

**Comment by ethanaobrien** [comment]
Let's go this route - I cant tell what you have and what you havent changed on your site because it doesnt include a changes page or even say this is the emulator you are using so for that I cant help you

**Comment by Vimm2** [comment]
> I'm not adding the button to the bottom, its better in the right click menu.

shaylevary has a valid point.  Mobile users don't have a mouse and thus can't right click, making that menu useless to them.
In addition, mobile users tend to have limited storage and thus need cache management the most.  An unlimited cache that can't be cleared is a problem.

**Comment by ethanaobrien** [comment]
That is true, I probably should put it down there with the next update I have planned. @shaylevary how good are you at creating svgs?

**Comment by shaylevary** [comment]
Hehe I'm not that good in drawing.
 But can you use something similar to this one in svg:
https://www.onlinewebfonts.com/icon/272

![img_272](https://user-images.githubusercontent.com/10597517/187559980-23805570-ec97-4a15-97b1-fa0b81b1f990.png)

Btw, I would like you to think again about adding 'clear all' button inside the cache manager, because if I got 90 rom files in archive, I will get tired of clicking each of the 50 rom files in order to empty my cache.

**Comment by ethanaobrien** [comment]
I'll look into adding a cache all

I actually really like that logo, I'll look into it and use something like it

Thanks!
